WL-4190: Add Export button for a reading list

nicholaswilson100 opened the pull request:
I removed the 'export all' option in the dropdown when doing ORLIMS so this is a new 'Export' button next to the 'Import' button to replace that functionality.

It only shows when there are citations in the reading list and not when you first create a list or haven't added any citations yet. 

The export dropdown option doesn't currently work on nightly because of permission issues:
e.g., NetworkError: 403 Forbidden - https://trunk-mysql.nightly.sakaiproject.org/access/citation/export_ris_all/6976cca7-7c00-4992-811f-5e324c615d04?citationCollectionId=6976cca7-7c00-4992-811f-5e324c615d04&resourceDisplayName=" -  this is caused by the fact that a null resourceId is being passed to CitationListAccessServlet.handleExportRequest which does a permissions check and a 'null' value fails that permissions check.  This fix is to pass 'contentCollectionId' instead of 'resourceId' to the check because even though 'citationCollectionId' is actually what is used in the Java to generate the exported file, there is a one to one mapping between a contentCollection and a citationCollection.  Moreover, feeding in the correct 'resourceId' value fails the permissions check anyway.

Individual export links, which we also want, are in a separate Jira WL-4196

nicholaswilson100 committed:
WL-4190: Add Export button for a reading list

buckett commented:
So what's to stop me from feeding in a contentCollectionId of a reading list in MyWorkspace and the citationCollectionId from reading list in a site I don't have access to?

nicholaswilson100 committed:
WL-4196: Export links for each citation on Edit Reading list page

WL-4190: Do permissions check on resourceId

WL-4190: Use resourceId to get citationCollectionId to avoid security bypasses

WL-4190: Fix indentation

WL-4190: use instance rather than cover

WL-4190: Changed static call to instance variable

WL-4190: Update import and make variable have method scope

nicholaswilson100 commented:
Good point.  

I see this Jira: https://jira.sakaiproject.org/browse/SAK-21875.  Is this the best one to patch (there are some related others and the comments are mixed )...

nicholaswilson100 commented:
Actually looking at the commit for this Jira, it looks like we have a lot of the code anyway and (like your Jira comment says) it's not a great fix anyway.  

Presumably there needs to be permissions check on the user that they have 'export' or 'read' rights on the list.  Sound right to you?

nicholaswilson100 commented:
I've piggybacked WL-4196 on this request as it uses some of the code here.

buckett commented:
Yeah, doing the permission check and the export on the same ID seems right.

nicholaswilson100 commented:
Ok so I'm thinking of doing the same permissions check for Export a reading list that is done when someone clicks on a item in Resources to view it?

nicholaswilson100 commented:
Ok so the permissions check for an exported citation or citation list is now against its 'resourceId'.

buckett commented:
I don't see any changes to the permission checking in this pull request.

nicholaswilson100 commented:
The combination of CitationHelperAction.java line 2670 which adds the resourceId to the Export link parameters and CitationListAccessServlet.java line 134 which checks the user has read rights on the resourceId is the permission check.

nicholaswilson100 commented:
We're now getting the citationCollectionId from the resourceId and no longer feeding in the citationCollectionId.

buckett commented on citations-impl/impl/src/java/org/sakaiproject/citation/impl/CitationListAccessServlet.java:
```diff
@@ -149,7 +150,25 @@ protected void handleExportRequest(HttpServletRequest req, HttpServletResponse r
+				citationCollectionId = new String(resource.getContent());
+			}
+			catch (PermissionException e) {
+				throw new EntityPermissionException(SessionManager.getCurrentSessionUserId(), "handleExportRequest", ref.getReference());
```
It's much better to use the ComponentManager to get an instance of a service as the static covers should be going away.

nicholaswilson100 commented:
We're now using service instance rather than cover

buckett commented:
I was actually meaning for the new dependency (SessionManager) which was added.

nicholaswilson100 commented:
Changed static call to instance variable for SessionManager

buckett commented on citations-impl/impl/src/java/org/sakaiproject/citation/impl/CitationListAccessServlet.java:
```diff
@@ -149,7 +150,27 @@ protected void handleExportRequest(HttpServletRequest req, HttpServletResponse r
+				citationCollectionId = new String(resource.getContent());
+			}
+			catch (PermissionException e) {
+				org.sakaiproject.tool.api.SessionManager sessionManager = (org.sakaiproject.tool.api.SessionManager) ComponentManager.get(org.sakaiproject.tool.api.SessionManager.class);
```
You can update the import. Rather than fully qualifying it.

A nice pattern to use is to have a member variable for the service and only do the lookup on `init()` of the servlet.

nicholaswilson100 replied:
That's not possible here. It's not a servlet so there is no init() method to override. It's called CitationListAccesServlet but it implements an interface HttpAccess rather than extending HttpServlet (or something similar).

nicholaswilson100 replied:
It is possible to botch something together to make it act like a servlet more but that feels like a hack to me and the code is already quite messy.

nicholaswilson100 commented:
I've updated the import and given the 2 service variables method scope.
